bn_mp_to_radix.c

```diff
@@ -5,66 +5,147 @@
 
 /* stores a bignum as a ASCII string in a given radix (2..64)
  *
- * Stores upto maxlen-1 chars and always a NULL byte
+ * Stores upto "size - 1" chars and always a NULL byte, puts the number of characters
+ * written, including the '\0', in "written".
+ * For "size == 0u" the whole number in "a" will get converted.
  */
-mp_err mp_to_radix(const mp_int *a, char *str, size_t maxlen, int radix)
+mp_err mp_to_radix(const mp_int *a, char **str, size_t size, size_t *written, int radix)
 {
-   int     digs;
+   size_t digs;
    mp_err  err;
    mp_int  t;
    mp_digit d;
-   char   *_s = str;
+   char   *_s, *tmp;
+   size_t len;
 
-   /* check range of the maxlen, radix */
-   if ((maxlen < 2u) || (radix < 2) || (radix > 64)) {
+   /* check range of radix and size*/
+   /* "size" includes the NUL byte so must be at least two if it isn't zero which
+       starts the automatic length-evaluation */
+   if ((size == 1u) || (radix < 2) || (radix > 64)) {
       return MP_VAL;
    }
-
    /* quick out if its zero */
    if (MP_IS_ZERO(a)) {
-      *str++ = '0';
-      *str = '\0';
+      if (size == 0u) {
+         if (*str != NULL) {
+            /* No MP_REALLOC because the old size is unknown */
+            tmp = realloc(*str, 2u);
+            if (tmp == NULL) {
+               return MP_MEM;
+            }
+            *str = tmp;
+         } else {
+            *str = MP_MALLOC(2u);
+            if (*str == NULL) {
+               return MP_MEM;
+            }
+         }
+      }
+      **str = '0';
+      (*str)++;
+      **str = '\0';
+      if (written != NULL) {
+         *written = 2u;
+      }
```

Comment:
Looks overly complex. No realloc since the user could pass static array. size==0&&*str!=0 and vice versa should be an error.

Reply:
> No realloc since the user could pass static array. size==0&&*str!=0 and vice versa should be an error.

I'd say either we do the alloc or the user does the alloc, I'm no fan of having both as an option through the same API